core/src/kiso.rs

```diff
@@ -0,0 +1,143 @@
+//! no any part of Iroha is interested in the whole state. However, the API could be extended
+//! in future.
+//!
+//! ## Example
+//!
+//! ```
+//! #[tokio::main]
+//! async fn main() {
+//!   todo!()
+//! }
+//! ```
+
+use eyre::Result;
+use iroha_config::{
+    client_api::{ConfigurationDTO, Logger as LoggerDTO},
+    iroha::Configuration,
+};
+use iroha_logger::actor::{Error as LoggerError, LoggerHandle};
+use tokio::sync::{mpsc, oneshot};
+
+const DEFAULT_CHANNEL_SIZE: usize = 32;
+
+/// The handle to work with the actor.
+///
+/// The actor will shutdown when all its handles are dropped.
+#[derive(Clone)]
+pub struct KisoHandle {
+    sender: mpsc::Sender<Message>,
+}
+
+impl KisoHandle {
+    /// Spawn a new actor
+    pub fn new(state: Configuration, logger: LoggerHandle) -> Self {
+        let (sender, receiver) = mpsc::channel(DEFAULT_CHANNEL_SIZE);
+        let mut actor = Actor {
+            receiver,
+            state,
+            logger,
+        };
+        tokio::spawn(async move { actor.run().await });
+
+        Self { sender }
+    }
+
+    /// Fetch the [`ConfigurationDTO`] from the actor's state.
+    ///
+    /// # Errors
+    /// If communication with actor fails.
+    pub async fn get_dto(&self) -> Result<ConfigurationDTO, Error> {
+        let (send, recv) = oneshot::channel();
+        let msg = Message::GetDTO { respond_to: send };
+        let _ = self.sender.send(msg).await;
+        let dto = recv.await?;
+        Ok(dto)
+    }
+
+    /// Update the configuration state, applying side effects.
+    ///
+    /// Awaits until the update is applied.
+    ///
+    /// # Errors
+    /// - If updating failure occurs
+    /// - If communication with actor is failed
+    pub async fn update_with_dto(&self, dto: ConfigurationDTO) -> Result<(), Error> {
+        let (send, recv) = oneshot::channel();
+        let msg = Message::UpdateWithDTO {
+            dto,
+            respond_to: send,
+        };
+        let _ = self.sender.send(msg).await;
+        recv.await?
+    }
+}
+
+enum Message {
+    GetDTO {
+        respond_to: oneshot::Sender<ConfigurationDTO>,
+    },
+    UpdateWithDTO {
+        dto: ConfigurationDTO,
+        respond_to: oneshot::Sender<Result<(), Error>>,
+    },
+}
+
+/// TODO
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// TODO
+    #[error("cannot dynamically update the configuration")]
+    Update(#[from] iroha_logger::ReloadError),
+    /// TODO
+    #[error("cannot get actor's response")]
+    Communication(#[from] oneshot::error::RecvError),
+}
+
+impl From<LoggerError> for Error {
+    fn from(value: LoggerError) -> Self {
+        match value {
+            LoggerError::LevelReload(err) => Self::from(err),
+            LoggerError::Communication(err) => Self::from(err),
+        }
+    }
+}
+
+struct Actor {
+    receiver: mpsc::Receiver<Message>,
+    state: Configuration,
+    logger: LoggerHandle,
```

Comment:
Just an idea: if we would add dynamic configuration for other more general-purpuse actors, we _might_ want to use a separate "handle" type not to conflate the cfg-related commands/messages with the actual actor functionality

Reply:
Hmm, I can't get wdym. Could you give an example?

Reply:
Hmm, actually now I see a way how to both use `watch` to propagate changes, but also update the actual config value in Kiso only when the change is applied. 

```mermaid
classDiagram
  class Kiso {
    Config config
    watch::Sender~LogLevelUpdate~ log_level_broadcast
    log_level_watcher() watch::Receiver~LogLevelUpdate~
  }
  class LogLevelUpdate {
    Level value
    oneshot::Sender~none~ respond_to
  }
  class Logger {
    watch_log_level(watch::Receiver~LogLevelUpdate~)
  }

  Logger <-- Kiso: sets log level
  Logger --> Kiso: notifies that log level is updated
```

```rust
let kiso: KisoHandle = _;
let logger: LoggerHandle = _;

logger.watch_log_level(kiso.log_level_watcher())
```

_Not sure how much this diagram helps_



My idea is that receiving part of `watch` channels might receive messages with `oneshot` channels to which they might reply when the change is actually applied. Therefore, when `Kiso` receives that update, it might update its `config` value, so that `GET /configuration` requests before actual value application will not return inaccurate information.

Anyway, this seem to make the whole system a little more complicated...

Reply:
Yeah, if we don't want it to be "fire and forget", it would probably should work like this. Though, in this case, I think, watch channels don't bring much to the table and can be just a separate `mpsc`.

Fire-and-forget or not, what I want is separation of for the actors: I think configuration updates should not go into the "main" channel, as it would require adding additional enum variants to each actor's messages and instead be handled via a separate and reusable channel type.

Though this change only makes sense if we are planning to add more updatable configuration options, otherwise it is indeed an unnecessary complication.

Reply:
Indeed, sounds convincing. Since implementing atomicity and transaction semantics would be way too complex for a not-so-important subsystem of Iroha, fire-and-forget would be a good fit.

Let me refactor.

Reply:
Refactored.

My main goal was to loose modules coupling. Now Iroha spawns another task on startup which is responsible for updates broadcasting. It listens for `watch` updates from Kiso and forwards them to Logger, without them knowing about each other.

This seems to be the most flexible way in terms of system components independence. However, Kiso has some boilerplate which I think doesn't make sense to refactor now.